### PR TITLE
Update MarkDequantization transformation

### PR DESCRIPTION
**Pull request by itikhono**
### Details:

**Original issue description:**

> Inside MarkDequantizationSubgraph, we try to find the  dequantization subgraph and mark it with disable const folding and keep const precision attributes.
> The dequantization subgraph has a relaxed structure and allows "any_input" as the input of Convert operation.
> 
> from MarkDequantizationSubgraph:
> {A09C9C08-03C1-4A48-8EAD-60D7EF76A4B5}
> 
> in the current case this is
>  (any input: Constant->Reshape) -> (necessary Convert) 
> 
> Constant->Reshape will be ConstFolded. ConstFolding doesn't run data copying in case of Reshape operation, so this is valid scenario.
> 
> MarkDequantizationSubgraph transformation marks Reshape op with "KeepConstPrecision" attribute.
> But after ConstFolding, KeepConstPrecision attr won't be copied to resulting constant because of
> {FE3E166D-17BD-410F-A08F-4502FB9BB3D0}
> and the whole dequantization subgraph will be const folded.

**Changes:**

- MarkDequantizationSubgraph logic was split into 2 transformations:  MarkDequantization, KeepConstPrecision

### Tickets:
 - *CVS-156576*
 - *CVS-156329*

### CODE REVIEW

**Review by v-Golubev — COMMENTED** [review]
My main concern is that we start propagating the rt_info which shouldn't be propagated. I have an alternative idea on how the described issue can be fixed.

Currently, `MarkDequantizationSubgraph` does several markups including:

1. `disable_constant_folding` markup for converts
2. `enable_keep_const_precision` markup for constants
3. The rest markup (`mark_as_dequantization_node` or `unmark_as_decompression`)

What if we move `enable_keep_const_precision` markup in a separate matcher pass, and organize the markup pipeline in the following way?

1.  First pass will have the same matcher as `MarkDequantizationSubgraph` currently has, and will perform only `disable_constant_folding` markup.
2. Then, `ConstantFolding` will be called, and all the intermediate layers between weights and convert will be folded. After that, the subgraph will have canonical form (low precision weights -> Convert -> (optional)Subtract -> Multiply)
3. At the end, `enable_keep_const_precision` markup will match the subgraphs with canonical form and mark constant nodes in this pattern

What do you think?

**Comment by itikhono** [comment]
> My main concern is that we start propagating the rt_info which shouldn't be propagated. I have an alternative idea on how the described issue can be fixed.
> 
> Currently, `MarkDequantizationSubgraph` does several markups including:
> 
> 1. `disable_constant_folding` markup for converts
> 2. `enable_keep_const_precision` markup for constants
> 3. The rest markup (`mark_as_dequantization_node` or `unmark_as_decompression`)
> 
> What if we move `enable_keep_const_precision` markup in a separate matcher pass, and organize the markup pipeline in the following way?
> 
> 1. First pass will have the same matcher as `MarkDequantizationSubgraph` currently has, and will perform only `disable_constant_folding` markup.
> 2. Then, `ConstantFolding` will be called, and all the intermediate layers between weights and convert will be folded. After that, the subgraph will have canonical form (low precision weights -> Convert -> (optional)Subtract -> Multiply)
> 3. At the end, `enable_keep_const_precision` markup will match the subgraphs with canonical form and mark constant nodes in this pattern
> 
> What do you think?

Yes, it's possible. But wouldn't that complicate the already complicated pipeline?
As I can see, `KeepConstPrecision` is used only for `Constants` in `ConvertPrecision` transformation, the flag will be ignored in case of other ops marked. So, it's precisely unsafe to copy this attribute into `Constants`, not to other ops. 

e.g.
1. initial subgraph: Parameter -> Reshape( "target_shape" const as an input) -> Convert
2. Reshape is marked with KeepConstPrecision flag
3. some transformation replaces the Reshape: Const -> Reshape (new "target_shape" Const marked with KeepConstPrecision)

"target_shape" Const will keep the original precision and won't be affected by ConvertPrecision transformation.
This may cause some implicit issues.

I believe we can expect constant subgraphs to fold and the issues do not arise.
Do we have non-constant (Parameter dependent) subgraphs in the real scenarios in this pattern?
Can this flag somehow extend beyond the dequantization subgraph?

**Comment by v-Golubev** [comment]
> Do we have non-constant (Parameter dependent) subgraphs in the real scenarios in this pattern?

Theoretically, we can have the subgraphs, which are marked by `MarkDequantizationSubgraph`, on non-constant path. For example, LPT supports the following subgraphs coming from some frontends (a picture from [lpt docs](https://docs.openvino.ai/2024/documentation/openvino-extensibility/openvino-plugin-library/advanced-guides/low-precision-transformations.html)):
image

In this case, the dequantization subgraph has canonical form, and f32->u8 convert allows `MarkDequantizationSubgraph` to match by precision on this convert as `data` node. And such subgraphs are placed both on data flow and weights.

> Can this flag somehow extend beyond the dequantization subgraph?

It seems like rt info from convert ops is not propagated through the graph during LPT. However, LPT are not the single interaction point with such canonical subgraphs: decompression related passes can also fuse ops from the subgraph in a separate operation, e.g. `GatherCompressed`. On your branch, we get the following graph after the fusion:
image

So `keep_const_precision` attribute is placed on non-constant flow after the fusion. Currently, we have no transformations which could work with `GatherCompressed`, but there is no guarantee that such optimizations will not exist in the future, and in case of such unwanted rt_info propagation we could get the problems you described in the example.

In conclusion, we will most likely not get any problems from the copyable `KeepConstPrecision` attribute in the current pipeline, but such change lays the foundation for potential future problems

**Comment by itikhono** [comment]
> > Do we have non-constant (Parameter dependent) subgraphs in the real scenarios in this pattern?
> 
> Theoretically, we can have the subgraphs, which are marked by `MarkDequantizationSubgraph`, on non-constant path. For example, LPT supports the following subgraphs coming from some frontends (a picture from [lpt docs](https://docs.openvino.ai/2024/documentation/openvino-extensibility/openvino-plugin-library/advanced-guides/low-precision-transformations.html)): image
> 
> In this case, the dequantization subgraph has canonical form, and f32->u8 convert allows `MarkDequantizationSubgraph` to match by precision on this convert as `data` node. And such subgraphs are placed both on data flow and weights.
> 
> > Can this flag somehow extend beyond the dequantization subgraph?
> 
> It seems like rt info from convert ops is not propagated through the graph during LPT. However, LPT are not the single interaction point with such canonical subgraphs: decompression related passes can also fuse ops from the subgraph in a separate operation, e.g. `GatherCompressed`. On your branch, we get the following graph after the fusion: image
> 
> So `keep_const_precision` attribute is placed on non-constant flow after the fusion. Currently, we have no transformations which could work with `GatherCompressed`, but there is no guarantee that such optimizations will not exist in the future, and in case of such unwanted rt_info propagation we could get the problems you described in the example.
> 
> In conclusion, we will most likely not get any problems from the copyable `KeepConstPrecision` attribute in the current pipeline, but such change lays the foundation for potential future problems

Ok, in this case, I will split the transformation into 2:
1. disable_constant_folding markup.
2. enable_keep_const_precision markup

**Comment by itikhono** [comment]
CPU functional test failure is not related to this PR.
GPU test failure is under investigation.

Additional perf validation will be triggered for this PR.

File: src/core/src/pattern/op/or.cpp [hunk 1]
```diff
@@ -13,7 +13,7 @@ bool ov::pass::pattern::op::Or::match_value(Matcher* matcher,
         auto saved = matcher->start_match();
         if (matcher->match_value(input_value, graph_value)) {
             auto& pattern_map = matcher->get_pattern_value_map();
-            pattern_map[input_value.get_node_shared_ptr()] = graph_value;
+            pattern_map[shared_from_this()] = graph_value;
```
Comment (itikhono) [inline src/core/src/pattern/op/or.cpp:16]: todo: add test
This is another fix. probably it's better to move it in the separate PR.

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp [hunk 2]
```diff
@@ -291,10 +291,12 @@ void TransformationsPipeline::apply(std::shared_ptr<ov::Model> func) {
 
         auto is_model_quantized = ov::pass::low_precision::LowPrecision::isFunctionQuantized(func);
         enableInt8 = config.get_property(ov::intel_gpu::enable_lp_transformations) && is_model_quantized;
-        if (enableInt8) {
-            manager.register_pass<ov::pass::MarkDequantizationSubgraph>(
-                std::vector<ov::element::Type>{ ov::element::i8, ov::element::u8, ov::element::i4, ov::element::u4 });
-        }
+
+        //if (enableInt8) { Why do we need this check? According to the line 378 we did this marking anyway
```
Comment (itikhono) [inline src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp:295]: todo: clarify this with GPU team

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp [hunk 3]
```diff
@@ -129,25 +129,27 @@ bool ov::pass::MOCTransformations::run_on_model(const std::shared_ptr<ov::Model>
     manager.set_per_pass_validation(false);
     using namespace ov::pass;
     REGISTER_PASS(manager, InitNodeInfo)
-    if (m_low_precision_enabled) {
-        manager.register_pass<ov::pass::MarkDequantizationSubgraph>(
-            element::TypeVector{ov::element::i8, ov::element::u8, ov::element::i4, ov::element::u4});
-    }
+    REGISTER_PASS(manager, EliminateConvert)
```
Comment (itikhono) [inline src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:132]: todo: revert this change

**Review by v-Golubev — APPROVED** [review]
No comments left from my side (except minor existing ones). Good job 👍

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp (hunk 3 above)
Comment (CuriousPanCake) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp:132]: Does it still need to be done?

File: src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp [hunk 4]
```diff
@@ -5,114 +5,174 @@
...
-                ov::disable_constant_folding(scale);
-                ov::unmark_as_decompression(scale);
-                ov::enable_keep_const_precision(scale->get_input_node_shared_ptr(0));
-            }
+        set_rt_info(pt_map, disable_constant_folding, converts_to_mark, precisions);
+        set_rt_info(pt_map, enable_constant_folding, converts_to_unmark, precisions);
+
+        // Move Reshape/Unsqueeze ops up to fold them in ConstantFolding.
+        auto changed = swap_nodes(pt_map, zp_convert_pattern, zp_reshape_pattern);
+        changed = changed || swap_nodes(pt_map, scale_convert_pattern, scale_reshape_pattern);
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp:123]: swap_nodes() won't be called if changed is already true.

Comment (itikhono) [reply src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp]: fixed

File: src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp (hunk 3 above)
Comment (itikhono) [reply src/common/transformations/src/transformations/common_optimizations/moc_transformations.cpp]: done

File: src/common/low_precision_transformations/tests/mark_dequantization_subgraph_transformation.cpp [hunk 5]
```diff
@@ -11,10 +11,44 @@
 #include "transformations/rt_info/keep_const_precision.hpp"
 
 #include "common_test_utils/ov_test_utils.hpp"
+#include "transformations/convert_precision.hpp"
 
 using namespace ov;
 
-TEST_F(TransformationTestsF, MarkDequantizationSubgraphTransformation) {
+TEST_F(TransformationTestsF, KeepConstPrecision) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/low_precision_transformations/tests/mark_dequantization_subgraph_transformation.cpp:18]: done
but it looks like the rt_info comparison doesn't work, the same with other tests,  I will double check

File: src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp [hunk 6]
```diff
@@ -5,114 +5,172 @@
...
+    };
+
+    auto m = std::make_shared<Matcher>(multiply_pattern, "MarkDequantization");
+    this->register_matcher(m, callback);
+}
+
+ov::pass::KeepConstsPrecision::KeepConstsPrecision(const element::TypeVector& precisions,
+                                                   bool fold_subtract_const,
+                                                   bool fold_multiply_const) {
+    // data input:
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp:143]: done

File: src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp [hunk 7]
```diff
@@ -5,114 +5,172 @@
...
+
+using RTInfoSetter = std::function<void(const std::shared_ptr<ov::Node>& node)>;
+void set_rt_info(const PatternValueMap& pt_map,
+                 const RTInfoSetter& rt_info_setter,
+                 const NodeVector& pattern_nodes,
+                 const ov::element::TypeVector& precisions) {
+    for (const auto& pattern_node : pattern_nodes) {
+        if (pt_map.count(pattern_node)) {
+            auto node = pt_map.at(pattern_node).get_node_shared_ptr();
+            if (ov::as_type_ptr<v0::Convert>(node) && !check_precision(node->get_input_element_type(0), precisions)) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp:43]: done

File: src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp (hunk 2 above)
Comment (itikhono) [reply src/plugins/intel_gpu/src/plugin/transformations_pipeline.cpp]: need a confirmation from the GPU team

File: src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp [hunk 8]
```diff
@@ -5,114 +5,172 @@
...
+using namespace ov::pass::pattern;
+
+namespace {
+
+bool check_precision(const ov::element::Type_t type_to_check, const ov::element::TypeVector& precisions) {
+    return std::find(precisions.begin(), precisions.end(), type_to_check) != precisions.end();
+};
+
+using RTInfoSetter = std::function<void(const std::shared_ptr<ov::Node>& node)>;
+void set_rt_info(const PatternValueMap& pt_map,
```
[the remark this answers is not in the document]
Comment (itikhono) [reply src/common/transformations/src/transformations/low_precision/mark_dequantization_subgraph.cpp:33]: tried, but it decreases readability of the main transformation code, so I would prefer to leave as is

**Review by sshlyapn — APPROVED** [review]

**Review by pavel-esir — APPROVED** [review]
